### DISCUSSION TITLE
How can I load a custom .env.vercel file for my react app?

### DISCUSSION THREAD

**Discussion by EdmundMai** [vercel/vercel#8212] [Help]
In the past I had a custom command for my project for copying my `.env.vercel` file to `.env` before running `npm build`. However that doesn't seem to work anymore

I noticed this in the vercel code: https://github.com/vercel/vercel/blob/183e411f7c51ce22a03f2eb355a531ef3c3770bb/packages/cli/src/commands/build.ts#L172

So I tried renaming my `.env.vercel` file to `.env.preview.local`, however, it looks like it's still trying to load `.env`.

<img width="1085" alt="Mouse_Highlight_Overlay" src="https://user-images.githubusercontent.com/1857068/180325281-24220c1e-a6d8-43a8-8ba1-968c62073f1d.png">

What is the proper way of loading a custom .env file? My `.env` is used for development, not vercel

**Comment by wyardley** [comment]
Also, note that we have a ticket open for the custom build command being ignored. This happened recently, stopped, and then returned, and no response on our support ticket yet.

**Comment by amyegan** [comment]
Hi @EdmundMai and @wyardley. Thanks for reaching out, and sorry this problem is happening. I want to make sure I get this in front of the right team and that I have all the info they need to figure it out. I think I already know the answer to some of these, but don't want to assume.
- Is this happening with every build, or intermittently?
- Have you made any changes to `.env` files or build settings in the last few weeks?
- Are you willing/able to try using our UI to handle environment variables as a workaround?
  - https://vercel.com/docs/concepts/projects/environment-variables

**Comment by wyardley** [comment]
> Is this happening with every build, or intermittently?

It happened last week (on 7/12; consistently) and went back to normal sometime around 7/13. Then the problem came back around 7/19, and has been happening consistently since that time.

> Have you made any changes to .env files or build settings in the last few weeks?

No. Also, I tried adjusting the custom build command, and it was still ignored. Basically, we copy our `.env.vercel` file to `.env.production` before build in the custom command for the react build (which then pulls the "production" values). It's affecting multiple projects.

When this problem happens, we see the output clearly from the `vercel build` command (which I think is normally bypassed with the custom build command being honored?)
we see the text
```
> Loaded env from ".env"
```
which we don't see when this problem is not happening.

https://github.com/vercel/vercel/blob/183e411f7c51ce22a03f2eb355a531ef3c3770bb/packages/cli/src/commands/build.ts#L172

We also tried renaming `.env.vercel` to `.env.preview.local`, which, based on the code, we thought might get loaded over `.env`, but that didn't seem to work either.

> Are you willing/able to try using our UI to handle environment variables as a workaround?

I don't think that's practical here - there are quite a few, and it's across multiple projects. Just to be clear, we're not relying on _Vercel's_ loading of env vars, but only React's.

**Reply by amyegan** [reply]
Can you share what your last working deployment was so we can check the logs?

**Reply by EdmundMai** [reply]
Hi Amy, 
We have quite a few builds in between so it's hard to pinpoint the most recent working build.

Here's an ID of a deployment from a few days ago that was deployed properly: `H2yWBSm1fsfjSMPePbc2KDCXK5ym`

Let me know if you need any more info

**Reply by wyardley** [reply]
There are multiple projects and a lot of builds, and the build doesn't "fail" (it just points to the wrong env), so it's hard to see the working vs. non-working ones without looking at each build logs.
But _roughly_, around 15:26 Pacific on 0/7/19 is a working ("successful") build (`FsbDsZpBxxZBkG3AcMQZJXVzCPhx`)
And I see a non-working (bad) build on 07/19 at ~ 01:27 Pacific (`FitTkBBNwjZYbMUBNJEiaHDntjiY`)

If you dig up our support ticket, there should be more information that should allow you to request more information and / or get information about our organization, failed jobs, etc.

It's very clear in the job output, though... the "good" jobs don't display the `> Loaded env from ".env"` message.

**Reply by samcx** [reply]
@wyardley Can you try using using `&&` instead of a `;` in your Vercel **Build Command**?

**Reply by wyardley** [reply]
@samsisle thanks - testing that now.

**Reply by wyardley** [reply]
With that set, I am still seeming to see
```
Running "vercel build"
Vercel CLI 27.2.0
> Loaded env from ".env"
```

<img width="1057" alt="image" src="https://user-images.githubusercontent.com/10714426/180336620-f9dbe4b3-612b-40d5-af9c-c56e74844f61.png">

**Reply by samcx** [reply]
@wyardley Thank you for the update — we will be passing along this to our Engineers and getting back to you!

**Reply by wyardley** [reply]
Thanks. I did check and at some point, the custom build command may have gotten disabled on one project, but verified that a) it still happens with it re-enabled, and b) that it is happening on other projects of ours where the custom build command was still enabled.

LMK if there's anything I need to do to clear caches or refresh after updating the setting to have the `&&`, but so far, doesn't seem to have helped.

**Reply by wyardley** [reply]
As best we can tell, when things are "working", Vercel is following the custom build command, and `vercel build` is not executed, and when things are "not working", Vercel is completely ignoring the custom build command and running `vercel build`.

**Reply by samcx** [reply]
@wyardley This `Running "vercel build"` is part of our new build pipeline changes, and it's likely affecting your build here. Will be relaying to our Engineers and getting back to you.

**Reply by amyegan** [reply]
Hello again @wyardley and @EdmundMai. The Engineering team has identified a solution and is currently working to implement it.

**Reply by amyegan** [reply]
Hi @wyardley and @EdmundMai. The fix was released and the build is using the new code as of today. Please let me know if you're still seeing the problem with your custom Build Command and I'll continue working with you and the build engineers to get it solved.

**Reply by wyardley** [reply]
Hi - yes, it seems like things have resolved as of a little while back. Appreciate the responsiveness and resolution!
